text/0000-poll-drop-ready.md

```diff
@@ -0,0 +1,480 @@
+Ready cannot necessarily be manufactured more than once. Because
+`poll_drop_ready` evaluates to a Poll of unit, which is a zero sized type that
+is trivial to produce, it is much more straightforward to implement it with
+fused semantics.
+
+### Why `poll_drop_ready` does not use `Pin`
+
+It would have been ideal for `poll_drop_ready` to use `Pin`, just like `poll`
+does. However, this would be unsound because of the definition of the Drop
+trait. This is a reoccurrence of an inconvenience caused by the definition of
+Drop, as it relates to pin projections.
+
+As first context: the definition of Drop is in some sense "wrong." It would be
+preferable of `Drop::drop` took self as `Pin<&mut Self>`, just like
+`Future::poll` does. Once something is dropped, we know it won't be moved again
+outside of the destructor. But because Drop receives an unpinned mutable
+reference, we know that it *can* move self, or even a field of self.
+
+Moving *fields* of self is a particular problem because of how it interacts
+with pin *projections* - that is, taking a Pin reference to type and
+"projecting" to a pinned reference to a field of that type. If you move out of
+the fields of a struct in the struct's destructor, pin projecting to that field
+would be unsound, because it would have been moved after it was pinned and
+before its destructor ran.
+
+This applies equally well to `poll_drop_ready`. If `poll_drop_ready` took self
+by `Pin`, in order to generate drop glue it would need to pin project to each
+field of the type. The order of operations would be:
+
+1. `poll_drop_ready` runs, pin projecting to each field of the type.
+2. The type's destructor runs, potentially moving each field.
+3. Each field's destructor runs, after the type's destructor has potentially
+moved them.
+
+Because of this sequence, the in projection in `poll_drop_ready` would be
+unsound. Therefore, `poll_drop_ready` cannot take self by Pin. Here is a bit of
+code demonstrating what I mean:
+
+```rust
+// If `field` implements `poll_drop_ready` and `poll_drop_ready` used `Pin`,
+// `field` would be witnessed as pinned before calling drop on `Foo`.
+struct Foo<T: Default> {
+    field: T
+}
+
+impl<T: Default> Drop for Foo<T> {
+    fn drop(&mut self) {
+        // Then, in the destructor for `Foo`, we can move `field` and re-pin it
+        // at a different location, violating the guarantees of `Pin`.
+
+        let moved_field = mem::replace(&mut self.field, T::default());
+        let pinned: Box<Pin<T>> = Box::pin(moved_field);
+    }
+}
+```
+
+However, its worth noting that a user can locally reason about whether private
+fields can be moved or accessed again, based on the APIs exposed by a type,
+and, if necessary, reconstruct a pin of those fields. [The pin documentation
+contains relevant information.][pin-docs] In other words, if your drop
+implementation is implemented correctly, you can construct a `Pin` of self
+inside your `poll_drop_ready`.
```

Comment:
Does this now mean we can or can not call `async fn`s inside drop? I think in generally the answer is "no":
- We can not directly await since `poll_drop_ready` is not an async fn
- The workaround is: Create a `Future` through the async fn one intends to call, store it somewhere in a pinned location, and poll it on every call from inside `poll_drop_ready`
- The latter approach doesn't seem to work, since there is no `Pin` argument

If this is true this would be major limitation, since most of the ecosystem is now moving towards first class support for async functions. And one might e.g. not even be possible to write to an async channel anymore without support for that.

Reply:
We provide specific guarantees about how poll_drop_ready will be called, which one can use to prove the safety of calls `Pin::new_unchecked` in order to poll those futures.